File: 3.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int start = 0, end = 0;
        unordered_set<char> current;
        int max_length = 0;
        while (end < s.length()) {
            char current_char = s[end];
            while (current.count(current_char)) {
                current.erase(s[start++]);
            }
            current.insert(current_char);
            end ++;
            if (end - start > max_length) {
                max_length = end - start;
            }
        }
        return max_length;
    }
};

File: 3.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> occurCharacters = new HashSet<>();
        int begin = 0, end = 0;
        int maxLength = 0;
        while (end != s.length()) {
            while (occurCharacters.contains(s.charAt(end))) {
                occurCharacters.remove(s.charAt(begin));
                begin += 1;
            }
            occurCharacters.add(s.charAt(end));
            end += 1;
            maxLength = Math.max(maxLength, end - begin);
        }
        return maxLength;
    }
}